File: tests/verify/detach_drain.py
"""A real viewer sends every accepted pane byte before `detach`, waits for the server's exit
frame, and never sends bytes that followed the detach key.

The script plays the server side over a private socket so it can observe the exact frame order
the production viewer emits. A real session server supplies a genuine state frame to paint.
"""
import fcntl
import json
import os
from pathlib import Path
import pty
import select
import socket
import struct
import subprocess
import sys
import tempfile
import termios
import threading
import time

binary = str(Path(sys.argv[1]).resolve())


def send(peer, message):
    data = json.dumps(message).encode()
    peer.sendall(struct.pack('>I', len(data)) + data)


def receive(peer):
    def exact(size):
        data = b''
        while len(data) < size:
            chunk = peer.recv(size - len(data))
            assert chunk, 'viewer closed before the expected frame'
            data += chunk
        return data
    size = struct.unpack('>I', exact(4))[0]
    assert size <= 16 * 1024 * 1024
    return json.loads(exact(size))


with tempfile.TemporaryDirectory(prefix='fdrain-', dir='/tmp') as directory:
    root = Path(directory)
    env = os.environ.copy()
    env.update(HOME=directory, XDG_RUNTIME_DIR=directory,
               XDG_CONFIG_HOME=directory + '/config', XDG_STATE_HOME=directory + '/state',
               SHELL='/bin/sh', TERM='xterm-256color')
    server = subprocess.Popen([binary, 'serve', '--name', 'default'], env=env,
                              stdout=subprocess.DEVNULL, stderr=subprocess.PIPE)
    child = None
    master, slave = pty.openpty()
    stopped = threading.Event()

    def drain_terminal():
        while not stopped.is_set():
            if select.select([master], [], [], .05)[0]:
                try:
                    if not os.read(master, 65536):
                        return
                except OSError:
                    return
    drain = threading.Thread(target=drain_terminal)
    drain.start()
    try:
        source = root / 'fux/default.attach.sock'
        deadline = time.monotonic() + 5
        while not source.exists():
            assert server.poll() is None, 'isolated server exited'
            assert time.monotonic() < deadline, 'isolated server did not start'
            time.sleep(.02)
        with socket.socket(socket.AF_UNIX) as peer:
            peer.settimeout(5)
            peer.connect(str(source))
            send(peer, dict(type='hello', version=3, rows=12, columns=48))
            assert receive(peer) == {'hello': {'version': 3}}
            state = receive(peer)
            assert 'state' in state
        with socket.socket(socket.AF_UNIX) as listener:
            path = root / 'probe.sock'
            listener.bind(str(path))
            path.chmod(0o600)
            listener.listen(1)
            listener.settimeout(5)
            fcntl.ioctl(slave, termios.TIOCSWINSZ, struct.pack('HHHH', 12, 48, 0, 0))
            child = subprocess.Popen([binary, 'attach', '--socket', str(path)], env=env,
                                     stdin=slave, stdout=slave, stderr=slave)
            peer, _ = listener.accept()
            with peer:
                peer.settimeout(5)
                assert receive(peer)['type'] == 'hello'
                send(peer, {'hello': {'version': 3}})
                assert receive(peer)['type'] == 'resize'
                send(peer, state)
                os.write(master, b'PREVIOUS_CHUNK')
                data = b''
                while data != b'PREVIOUS_CHUNK':
                    message = receive(peer)
                    assert message['type'] == 'input', message
                    data += bytes(message['bytes'])
                # A separate read: detach followed by a command that must never be sent.
                os.write(master, b'\x01d\x01t')
                message = receive(peer)
                assert message == {'type': 'detach'}, message
                time.sleep(.2)
                assert child.poll() is None, 'viewer exited before the server acknowledged detach'
                peer.setblocking(False)
                try:
                    extra = peer.recv(1)
                except BlockingIOError:
                    extra = None
                assert extra is None, f'trailing bytes after detach: {extra!r}'
                peer.setblocking(True)
                send(peer, {'exited': {'code': None}})
                assert child.wait(timeout=5) == 0
                assert peer.recv(1) == b'', 'viewer kept the connection after exit'
        print('PASS separate-read detach sends preceding input, waits for exit and drops the suffix')
    finally:
        if child is not None and child.poll() is None:
            child.kill(); child.wait()
        stopped.set(); drain.join(timeout=1)
        os.close(master); os.close(slave)
        server.terminate()
        try:
            server.wait(timeout=10)
        except subprocess.TimeoutExpired:
            server.kill(); server.wait()
            raise

File: tools/xtask/src/scenarios/detach_drain.rs
//! Observe real viewer input ordering and acknowledgment through a private server.
use crate::support::{
    attachment::{connect, receive, send},
    local::{Root, until},
    process::{Guard, wait},
};
use anyhow::{Result, ensure};
use serde_json::json;
use std::{
    fs,
    io::Read,
    os::unix::{fs::PermissionsExt, net::UnixListener},
    path::Path,
    process::Stdio,
    sync::{
        Arc,
        atomic::{AtomicBool, Ordering},
    },
    time::Duration,
};

struct Drain {
    stop: Arc<AtomicBool>,
    thread: Option<std::thread::JoinHandle<Result<()>>>,
}
impl Drain {
    fn start(fd: std::os::fd::OwnedFd) -> Result<Self> {
        nix::fcntl::fcntl(
            &fd,
            nix::fcntl::FcntlArg::F_SETFL(nix::fcntl::OFlag::O_NONBLOCK),
        )?;
        let stop = Arc::new(AtomicBool::new(false));
        let stopping = stop.clone();
        let thread = std::thread::spawn(move || {
            let mut bytes = [0; 65536];
            while !stopping.load(Ordering::Relaxed) {
                match nix::unistd::read(&fd, &mut bytes) {
                    Ok(0) | Err(nix::errno::Errno::EIO) => break,
                    Ok(_) | Err(nix::errno::Errno::EAGAIN) => {}
                    Err(error) => return Err(error.into()),
                }
                std::thread::sleep(Duration::from_millis(10));
            }
            Ok(())
        });
        Ok(Self {
            stop,
            thread: Some(thread),
        })
    }
    fn finish(&mut self) -> Result<()> {
        self.stop.store(true, Ordering::Relaxed);
        if let Some(thread) = self.thread.take() {
            thread
                .join()
                .map_err(|_| anyhow::anyhow!("terminal drain panicked"))??;
        }
        Ok(())
    }
}
impl Drop for Drain {
    fn drop(&mut self) {
        if let Err(error) = self.finish() {
            eprintln!("terminal drain: {error:#}");
        }
    }
}

pub(super) fn run(binary: &Path) -> Result<()> {
    let root = Root::new("fdrain-rs-", &["/bin/sh".into()])?;
    let mut server = root.server(binary)?;
    let mut source = connect(&root.path().join("fux/default.attach.sock"))?;
    send(
        &mut source,
        &json!({"type":"hello","version":6,"rows":12,"columns":48}),
    )?;
    ensure!(
        receive(&mut source)? == json!({"hello":{"version":6}}),
        "source hello"
    );
    let state = receive(&mut source)?;
    ensure!(state.get("state").is_some(), "source state missing");
    drop(source);
    let path = root.path().join("probe.sock");
    let listener = UnixListener::bind(&path)?;
    fs::set_permissions(&path, fs::Permissions::from_mode(0o600))?;
    listener.set_nonblocking(true)?;
    let pty = crate::support::pty::open(12, 48)?;
    let mut drain = Drain::start(pty.master.try_clone()?)?;
    let mut child = Guard(
        root.command(binary)
            .args(["attach", "--socket"])
            .arg(&path)
            .stdin(Stdio::from(pty.slave.try_clone()?))
            .stdout(Stdio::from(pty.slave.try_clone()?))
            .stderr(Stdio::from(pty.slave.try_clone()?))
            .spawn()?,
    );
    let (mut peer, _) = until(Duration::from_secs(5), || match listener.accept() {
        Ok(value) => Ok(Some(value)),
        Err(error) if error.kind() == std::io::ErrorKind::WouldBlock => Ok(None),
        Err(error) => Err(error.into()),
    })?;
    peer.set_nonblocking(false)?;
    peer.set_read_timeout(Some(Duration::from_secs(5)))?;
    peer.set_write_timeout(Some(Duration::from_secs(5)))?;
    ensure!(receive(&mut peer)?["type"] == "hello", "viewer hello");
    send(&mut peer, &json!({"hello":{"version":6}}))?;
    ensure!(receive(&mut peer)?["type"] == "resize", "viewer resize");
    crate::support::attachment::send_server(&mut peer, &state)?;
    let previous = b"PREVIOUS_CHUNK";
    ensure!(
        nix::unistd::write(&pty.master, previous)? == previous.len(),
        "short input write"
    );
    let mut data = Vec::new();
    while data != previous {
        let message = receive(&mut peer)?;
        ensure!(message["type"] == "input", "wrong input frame: {message}");
        let chunk: Vec<u8> = serde_json::from_value(message["bytes"].clone())?;
        ensure!(!chunk.is_empty(), "empty input frame");
        data.extend(chunk);
        ensure!(previous.starts_with(&data), "input changed: {data:?}");
    }
    ensure!(
        nix::unistd::write(&pty.master, b"\x01d\x01t")? == 4,
        "short detach write"
    );
    ensure!(
        receive(&mut peer)? == json!({"type":"detach"}),
        "wrong detach frame"
    );
    std::thread::sleep(Duration::from_millis(200));
    ensure!(
        child.0.try_wait()?.is_none(),
        "viewer exited before detach acknowledgment"
    );
    peer.set_nonblocking(true)?;
    let mut byte = [0];
    ensure!(
        matches!(peer.read(&mut byte), Err(error) if error.kind() == std::io::ErrorKind::WouldBlock),
        "trailing bytes or EOF after detach"
    );
    peer.set_nonblocking(false)?;
    send(&mut peer, &json!({"exited":{"code":null}}))?;
    ensure!(
        wait(&mut child.0, Duration::from_secs(5))?.success(),
        "viewer exit failed"
    );
    ensure!(
        peer.read(&mut byte)? == 0,
        "viewer retained connection after exit"
    );
    drain.finish()?;
    server.finish()?;
    println!(
        "PASS separate-read detach sends preceding input, waits for exit and drops the suffix"
    );
    Ok(())
}
